tools/PI/DevHome.PI/Controls/AddToolControl.xaml.cs

```diff
@@ -79,77 +128,295 @@ private void OkButton_Click(object sender, RoutedEventArgs e)
+        {
+            // Update the progress report.
+            LoadingProgressTextRing.Value = percent;
+        });
+
+        await ProcessItemsAsync(progress);
+
+        foreach (var app in _allApps)
+        {
+            SortedApps.Add(app);
+        }
+
+        IsLoading = false;
+    }
+
+    private int GetShortcuts()
+    {
+        int count;
+
+        // Search for .lnk files in the per-user and all-users Start Menu Programs directories.
+        // %APPDATA%\Microsoft\Windows\Start Menu\Programs
+        var startMenuProgramsPath =
+            Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.StartMenu), "Programs");
+
+        // %ProgramData%\Microsoft\Windows\Start Menu\Programs
+        var commonStartMenuProgramsPath =
+            Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.CommonStartMenu), "Programs");
+
+        _shortcuts = [];
+        if (Directory.Exists(startMenuProgramsPath))
+        {
+            _shortcuts.AddRange(Directory.GetFiles(startMenuProgramsPath, "*.lnk", SearchOption.AllDirectories));
+        }
+
+        if (Directory.Exists(commonStartMenuProgramsPath))
+        {
+            _shortcuts.AddRange(Directory.GetFiles(commonStartMenuProgramsPath, "*.lnk", SearchOption.AllDirectories));
+        }
+
+        count = _shortcuts?.Count ?? 0;
+        return count;
+    }
+
+    private int GetPackages()
     {
-        UpdateSampleCommandline();
+        var count = 0;
+        var packageManager = new PackageManager();
+        _packages = packageManager.FindPackagesForUserWithPackageTypes(string.Empty, PackageTypes.Main).ToList();
+        if (_packages is not null)
+        {
+            count = _packages.Count;
+        }
+
+        return count;
     }
 
-    private void UpdateSampleCommandline(object sender, RoutedEventArgs e)
+    private async Task ProcessItemsAsync(IProgress<int> progress)
     {
-        UpdateSampleCommandline();
+        // Process all the shortcut files.
+        var currentCount = 0;
+        for (var i = 0; i < _shortcuts?.Count; i++)
+        {
+            await Task.Run(() => ProcessShortcut(_shortcuts[i]));
+
+            // Report progress.
+            currentCount++;
+            var percentComplete = (i + 1) * ShortcutProcessingEndIndex / _itemCount;
+            progress.Report(percentComplete);
+        }
+
+        for (var j = 0; j < _packages?.Count; j++)
+        {
+            await Task.Run(() => ProcessPackage(_packages[j]));
+
+            // Report progress.
+            currentCount++;
+            var percentComplete = ShortcutProcessingEndIndex + ((j + 1) * ShortcutProcessingEndIndex / _itemCount);
+            progress.Report(percentComplete);
+        }
+
+        _allApps = _allApps.OrderBy(app => app.Name).ToList();
+
+        // We get the icon data on the UI thread, because BitmapImages must be created on the UI thread.
+        for (var k = 0; k < _allApps.Count; k++)
```

Comment:
>for (var k = 0; k < _allApps.Count; k++) [](http://example.com/codeflow?start=8&length=40)

nit: is it possible to use foreach?

Reply:
I need the loop counter as part of the progress report calculation.

Reply:
ah yes makes sense